sympy/calculus/util.py

```diff
@@ -344,6 +346,22 @@ def periodicity(f, symbol, check=False):
         except NotImplementedError:
             pass
 
+    if isinstance(f, Abs):
+        arg = f.args[0]
+        new_f = arg
+        if isinstance(arg, (sec, csc, cos)):
+            new_f = sin(arg.args[0])
+        period = periodicity(new_f, symbol)
+        if period is not None and isinstance(new_f, sin):
+            test = Abs(new_f.subs(symbol, symbol + period/2)) == Abs(new_f)
```

Comment:
The logic is too subtle for me -- I can't easily tell if this is always going to be right.

Reply:
I think since we are checking that Abs(f(x+period/2)) == Abs(f(x)) it can't return a wrong result.If this condition is True **_period/2_** is it's period and if not then it's period is _**period**_(period of new_f)
What the current version does to determine the period of Abs(f(x)) is that it just determines the period of f(x). But it misses that there maybe a value T such that **f(x+T) = -f(x)** for all x. In that case `f(x+2*T) = -f(x+T) = f(x)` so periodicity(Abs(f(x)), x) returns 2*T ( period of f(x) ). That is the reason I check with period/2  ```test = Abs(new_f.subs(symbol, symbol + period/2)) == Abs(new_f)```. 
Recasting csc,sin,cos to sin is ok since if Abs(sin(g(x+T))) = Abs(sin(g(x))) then Abs(cos(g(x+T))) will be equal to Abs(cos(g(x))) the same is true for csc and sec

Reply:
OK, I see. Would you be willing to go with this edit that adds commentary and delays introduction of `new_f`?

```
if isinstance(f, Abs):
    arg = f.args[0]
    if isinstance(arg, (sec, csc, cos)):
        # all but cot and tan might have
        # a period that is half as large
        # so recast as sin
        arg = sin(arg.args[0])
    period = periodicity(arg, symbol)
    if period is not None and isinstance(arg, sin):
        # the argument of Abs was a trigonometric other than
        # cot or tan; test to see if the half-period
        # is valid. Abs(arg) has behavior equivalent to
        # orig_f, so use that for the test:
        new_f = Abs(arg)
        test = new_f.subs(symbol, symbol + period/2) == new_f
        if check and (not test):
            raise NotImplementedError(filldedent('''
                The period of the given function cannot be verified.
                Set check=False to obtain the value.'''))
        elif test:
            return period/2
        return period
```

Reply:
Why just don't check difference of `new_f.subs(symbol, symbol + period/2)` and `new_f` is multiple of `pi` like `(new_f.subs(symbol, symbol + period/2) - new_f) % pi == S.Zero`?

Reply:
@ylemkimon Please give me an example where (new_f.subs(symbol, symbol + period/2) - new_f) is not equal to zero and divisible by pi

Reply:
Abs(cos(x)) would give period 2*pi and then difference is pi. A difference of zero would also meet the condition since zero is divisble by every number(except zero).

Reply:
if the f=Abs(cos(x)) then arg will be cos(x) ( **f.args[0]** ) and new_f will be Abs(sin(x)) after recasting.new_f will be a trigonometric function always. I think you are suggesting to again apply new_f.args[0]

Reply:
@souravghosh97 If you're commenting to me, yes. `new_f` in previous comments means argument of trigonometric function inside Abs.

Reply:
new_f itself can be trigonometric consider cos(sin(x))